lib/model/model.go

```diff
@@ -1334,26 +1362,38 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 	}
 
 	if deviceID != protocol.LocalDeviceID {
-		l.Debugf("%v REQ(in): %s: %q / %q o=%d s=%d t=%v", m, deviceID, folder, name, offset, len(buf), fromTemporary)
+		l.Debugf("%v REQ(in): %s: %q / %q o=%d s=%d t=%v", m, deviceID, folder, name, offset, size, fromTemporary)
 	}
 
-	folderFs := folderCfg.Filesystem()
-
 	if fs.IsInternal(name) {
-		l.Debugf("%v REQ(in) for internal file: %s: %q / %q o=%d s=%d", m, deviceID, folder, name, offset, len(buf))
+		l.Debugf("%v REQ(in) for internal file: %s: %q / %q o=%d s=%d", m, deviceID, folder, name, offset, size)
 		return protocol.ErrNoSuchFile
 	}
 
 	if folderIgnores.Match(name).IsIgnored() {
-		l.Debugf("%v REQ(in) for ignored file: %s: %q / %q o=%d s=%d", m, deviceID, folder, name, offset, len(buf))
+		l.Debugf("%v REQ(in) for ignored file: %s: %q / %q o=%d s=%d", m, deviceID, folder, name, offset, size)
 		return protocol.ErrNoSuchFile
 	}
 
 	if err := osutil.TraversesSymlink(folderFs, filepath.Dir(name)); err != nil {
-		l.Debugf("%v REQ(in) traversal check: %s - %s: %q / %q o=%d s=%d", m, err, deviceID, folder, name, offset, len(buf))
+		l.Debugf("%v REQ(in) traversal check: %s - %s: %q / %q o=%d s=%d", m, err, deviceID, folder, name, offset, size)
 		return protocol.ErrNoSuchFile
 	}
 
+	return nil
+}
+
+func (m *Model) request(conn protocol.Connection, res protocol.RequestResult, folderCfg config.FolderConfiguration, deviceID protocol.DeviceID, folder, name string, size int32, offset int64, hash []byte, weakHash uint32, fromTemporary bool) {
+	defer func() {
+		conn.Response(res)
+		<-res.Done
```

Comment:
I don't know what you're trying to do with Done here.

Apart from that, the approach seems sensible. The other way that strikes me is that the request method could return a future of sorts (`struct { Data <-chan []byte, Error <-chan error }` perhaps) that the protocol layer would launch a goroutine to wait on and do the actual response message creation, handling the request ID etc. It's not super clean, imho, that the model layer needs to know about and handle this.

Reply:
I think Done is a future. It uses a mechanism that was already used in the protocol package, that notifies the creator of the object (here 'protocol.RequestResult`) when it is finished handling the contained data (i.e. the taken bytes can be given back to the semaphore).

Together with the actual limiting that I added, does this make sense now?